vm/devices/storage/scsidisk/src/lib.rs

```diff
@@ -867,6 +885,134 @@ impl SimpleScsiDisk {
         Ok(RequestParameters { tx, offset, fua })
     }
 
+    fn validate_data_cdb16_atomic(
+        &self,
+        external_data: &RequestBuffers<'_>,
+        request: &Request,
+        sector_count: u64,
+    ) -> Result<RequestParameters, ScsiError> {
+        let cdb = scsi::Cdb16Atomic::read_from_prefix(&request.cdb[..])
+            .unwrap()
+            .0; // TODO: zerocopy: use-rest-of-range (https://github.com/microsoft/openvmm/issues/759)
+        let len = cdb.transfer_blocks.get() as u64;
+        let offset = cdb.logical_block.get();
+        let sector_shift = self.sector_shift;
+        let max = external_data.len() >> sector_shift;
+        let atomic_boundary = cdb.atomic_boundary.get();
+        if len == 0 || len as usize > max {
+            tracelimit::error_ratelimited!(len, max, "illegal block");
+            return Err(ScsiError::IllegalRequest(
+                AdditionalSenseCode::ILLEGAL_BLOCK,
+            ));
+        }
+
+        if sector_count <= offset || sector_count - offset < len {
+            tracelimit::error_ratelimited!(sector_count, offset, len, "illegal block");
+            return Err(ScsiError::IllegalRequest(
+                AdditionalSenseCode::ILLEGAL_BLOCK,
+            ));
+        }
+
+        // Validate atomic-specific parameters
+        // If atomic boundary is 0, make sure transfer length not larger than max atomic
+        // transfer length
+        if atomic_boundary == 0
+            && len as u32
+                > self
+                    .scsi_parameters
+                    .maximum_atomic_transfer_length
+                    .unwrap_or(0)
```

Comment:
If None is equivalent to 0, store it that way so that your code isn't littered with unwraps.